File: 0733-flood-fill/0733-flood-fill.py
class Solution:
    def floodFill(self, image: List[List[int]], sr: int, sc: int, color: int) -> List[List[int]]:
        visited = set()
        original = image[sr][sc]
        
        def travel(x, y):
            if (x, y) in visited: return
            if x < 0 or x >= len(image) or y < 0 or y >= len(image[0]): return
            if image[x][y] != original:return
            
            visited.add((x, y))
            image[x][y] = color
            
            travel(x + 1, y)
            travel(x - 1, y)
            travel(x, y + 1)
            travel(x, y - 1)
        
        travel(sr, sc)
        return image

File: 0733-flood-fill/0733-flood-fill.js
/**
 * @param {number[][]} image
 * @param {number} sr
 * @param {number} sc
 * @param {number} color
 * @return {number[][]}
 */
var floodFill = function(image, sr, sc, color) {
    const isInScope = (r, c) => r >= 0 && r < image.length && c >= 0 && c < image[0].length
    const originalColor = image[sr][sc]
    const isOriginalColor = (r, c) => originalColor === image?.[r]?.[c]
    const seen = new Set()
    
    const search = (r, c) => {
        const str = `${r}, ${c}`
        if (seen.has(str)) return
        if (!(isInScope(r, c) && isOriginalColor(r, c))) return
        image[r][c] = color
        seen.add(str)
        search(r - 1, c)
        search(r + 1, c)
        search(r, c - 1)
        search(r, c + 1)
    }
    
    if (!(originalColor === color)) search(sr, sc)
    return image
};